### PR TITLE
Ximea camera support

### PR DESCRIPTION
**Pull request by iandobbie**
Adds the basic ximea camera support which works in python 3.

### CODE REVIEW

File: microscope/cameras/ximea.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    def get_cycle_time(self):
+        return (self.handle.get_exposure()*1.0E-6)
+
+    def _get_sensor_shape(self):
+        return (self.img.width,self.image.height)
+
+
+    def get_trigger_type(self):
+        trig=self.handle.get_trigger_source()
+        return devices.TRIGGER_SOFT
```
Comment (mickp) [inline microscope/cameras/ximea.py]: This always returns TRIGGER_SOFT, regardless of the actual trigger setting. 
Does get_trigger_source return one of the XI_TRG_ constants, defined in the comments above? If so, these should be mapped to trigger modes/types that microscope recognises. I think David added a new mix-in for this, maybe?

File: microscope/cameras/ximea.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+    def _get_binning(self):
+         return (1,1)
+
+    @keep_acquiring
+    def _set_binning(self, h, v):
+        return False
+
+    def _get_roi(self):
+        return (0, 0, 512, 512)
```
Comment (mickp) [inline microscope/cameras/ximea.py]: If ROI and binning are not implemented, this should return the sensor size, not 512 x 512.

File: microscope/cameras/ximea.py (hunk 1 above)
Comment (mickp) [reply microscope/cameras/ximea.py]: I found the TriggerTargetMixIn, but could use some advice on the intended usage.

Comment (carandraug) [reply microscope/cameras/ximea.py]: > I found the TriggerTargetMixIn, but could use some advice on the intended usage.

One needs to set the private `_trigger_type` and `_trigger_mode` properties during construction. The value in this properties should be an instance of the `TriggerType` and `TriggerMode` enums. Then one needs to define the `set_trigger` abstract method with whatever implementation is suitable to the device.

Comment (iandobbie) [reply microscope/cameras/ximea.py]: I think this was a quick hack that got left in. The hardware trigger code does work and is being used on DeepSIM. This function should return a proper value of course!

File: microscope/cameras/ximea.py (hunk 2 above)
Comment (iandobbie) [reply microscope/cameras/ximea.py]: Yes, obviously a terrible hack, Shows that this code is no where near complete. Sorry.

File: microscope/cameras/ximea.py (hunk 1 above)
Comment (mickp) [reply microscope/cameras/ximea.py]: OK - so does set_trigger have to set the trigger on the hardware, and set _trigger_type and _trigger_mode appropriately? If so, couldn't we just pass a mapping to the constructor?

Comment (carandraug) [reply microscope/cameras/ximea.py]: > so does set_trigger have to set the trigger on the hardware, and set _trigger_type and _trigger_mode appropriately? If so, couldn't we just pass a mapping to the constructor?

Not necessarily setting on the hardware. The alpao mirror does not have a setting on the hardware to trigger mode. One needs to remember what mode was selected and code appropriately when queueing patterns.

I'm guessing with mapping, you mean a map from microscope trigger enum to the sdk trigger enum.  However, the sdk may support them without enums which is what happens in the alpao mirror. And some triggers triggers may support only a subset of trigger modes.

Basically, implementation is completely left up to the concrete class because of so many special cases.

Comment (mickp) [reply microscope/cameras/ximea.py]: Right - good points. Got it, I think. Where there is a clear n-to-1 relationship between device and microscope trigger modes, then we can just put a mapping on the device class, I guess.

File: microscope/cameras/ximea.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    def set_exposure_time(self, value):
+        #exposure times are set in us.
+        self.handle.set_exposure(int(value*1.0E6))
+
+    def get_exposure_time(self):
+        #exposure times are in us, so multiple by 1E-6 to get seconds.
+        return (self.handle.get_exposure()*1.0E-6) 
+
+    def get_cycle_time(self):
+        return (self.handle.get_exposure()*1.0E-6)
```
[the remark this answers is not in the document]
Comment (mickp) [reply microscope/cameras/ximea.py]: Good catch - missed that.

File: microscope/cameras/ximea.py (hunk 2 above)
Comment (mickp) [reply microscope/cameras/ximea.py]: Fixed by f2680e33b937deee75d7225105afe017408900ec

File: microscope/cameras/ximea.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (iandobbie) [reply microscope/cameras/ximea.py]: Fixed by calling get framerate to return real cycle time
